### DISCUSSION TITLE
How can I increase the number of cores used during assembly?

### DISCUSSION THREAD

**Discussion by mars1211** [cross-rs/cross#1225] [Q&A]
Hello everyone. 
I run the compilation "cross build --target x86_64-unknown-linux-gnu" and wait about 7.30 minutes for it to finish. If you compile everything together, it takes about 30 minutes. I opened the task manager and saw that several cores were being used, the rest were idle. 
I use Docker with WSL2 on core  i7-12700 12 cores 20 threads 4.9Ghz. When I run the build on WSL, all resources are used.
What needs to be done so that all cores are involved?

**Comment by Emilgardis** [comment]
in `$USERPROFILE/.wslconfig` you can set the limits for wsl which should be what docker CE uses

here's my current setup on my windows machine

```
❯ cat $env:USERPROFILE\.wslconfig
[wsl2]
memory=8GB
processors=12 
```

and the result when doing `cross build --target x86_64-unknown-linux-gnu`
![image](https://user-images.githubusercontent.com/1502855/226704985-7b11e96b-e0ad-4c0c-a578-4683df31b37b.png)


I'm not sure why you're seeing different results when using docker and wsl, but they should be the same.

If you want to see stats about the build, like % of cpu used, use `docker stats`

If that doesn't help, you might not be using wsl2 for docker or something else is wrong

**Reply by Emilgardis** [reply]
Also to note, it might be that this is how long it takes (I doubt it though, 30 minutes is a lot).

Do note that cargo/rustc doesn't always use all cores.

One thing that can slow things down is filesystem access. WSL2 <-> Windows is kinda slow, see https://stackoverflow.com/questions/62154016/docker-on-wsl2-very-slow

**Reply by mars1211** [reply]
I think my answer below will also be useful for you, judging by the CPU load.

**Comment by mars1211** [comment]
I found the problem. It consisted in the fact that the project uses files hosted in Windows, so the delays were due to the file system driver in WSL. 
I assembled the project in WSL from a mounted section, for example "/mnt/c/Projects/my-project" - it was going to be like in "cross" for about 7 minutes. Then I copied the project and built it from the home directory, for example "~/my-project" - the build took about 40 seconds.
A wonderful article prompted me to this idea - https://markentier.tech/posts/2022/01/speedy-rust-builds-under-wsl2/

**I think the following steps need to be added for this project:**
* copying the project folder to the virtual machine
* compilation
* copy back

For example, here is the CPU load with the command "cross build --target x86_64-unknown-linux-gnu" or "cargo build" from under the mounted disk: >20 minutes
![2023-03-22_15-41-46](https://user-images.githubusercontent.com/104355725/226919639-635a6fb7-794f-4ecb-b986-dbe71e0f5f02.png)

And so from under the copied project: 1 minute, 34 seconds
![image](https://user-images.githubusercontent.com/104355725/226919352-e6219a0e-79ea-4ef8-a464-799f95dbec41.png)
